### ISSUE TITLE
Range should support Infinite Ranges

### ISSUE THREAD

**Issue by jeffreyr-porch** [tminglei/slick-pg#234]
I'm trying to parse a daterange column out of a postgres database, and the data makes use of infinite ranges. The slick-pg Range support crashes on a null exception when this occurs (specifically, it is while calling toSQLDate from.

According to the Posgres Documentation [Section 8.17.4](http://www.postgresql.org/docs/9.3/static/rangetypes.html#RANGETYPES-INFINITE), postgres ranges can have a NULL for any combination of lower and upper bound, to symbolize an infinite range.

It would be desireable for the Range case class to declare the start and end values not as T, but as Option[T], setting them to None when the value, for example:
```scala
case class Range[T](start: Option[T], end: Option[T], edge: EdgeType = `[_,_)`) {
  ...
}
```

At the moment, I have tried several strategies for solving this, and in the end I wound up duplicating the PGRangeSupport.scala file into my project, renaming the class from Range[T] to InfRange[T], and editing the relevant code to support start/end being Option types. It is, obviously, an ugly solution, but less invasive attempts (including modifications to allow null in the existing Range type) proved problematic.

I'm hoping I missed something stupid, and that there is an easier way to do this.

**Comment by tminglei** [comment]
Hi @jeffreyr-porch, can you help provide some sample cases? I didn't notice this before.

**Comment by tminglei** [comment]
@jeffreyr-porch, thanks for pointing it out.
I also haven't a perfect solution yet, and will continue thinking.

**Comment by jeffreyr-porch** [comment]
It seems like the hardest part is backwards compatibility: One would like to maintain the existing Range case class for those that are already using it, but have a different Range case class with Option types to handle unbounded ranges.

It also occurs to me that in making my own version  of PGRangeSupport.scala with a different range type name, I don't know what (if any) surprises might come back to get me later.

Good luck, and let me know if you need/want anything from me.

**Comment by tminglei** [comment]
@jeffreyr-porch, I submitted a fixing patch, and seems it kept most of backwards compatibility.
You can help check it.

_Thanks! :-)_

**Comment by jeffreyr-porch** [comment]
Awesome! I'm heading out on vacation for the remainder of the week, but I'll see about trying to get this tested on my end next week.

**Comment by jeffreyr-porch** [comment]
Sorry for the couple month delay there. I have installed the source into a Slick 3.1 project I'm working on, and it seems to be working correctly for the use cases I tried.

Any idea when this patch would make it into the mainline? We have several teams that are poised to run into this problem, and I'd like to offer them a better solution than either copy/paste a file or maintaining our own patched version of slick-pg. :)

**Comment by tminglei** [comment]
Merged, maybe release it in these days. :-)

**Comment by jeffreyr-porch** [comment]
Huzzah! Thanks. :)

**Comment by tminglei** [comment]
@jeffreyr-porch just released slick-pg 0.11.1, pls check.

**Comment by jeffreyr-porch** [comment]
Thanks!

We did run into one bug:
```scala
def nextDateRangeOption() = r.nextStringOption().map(toSQLDate)
```
should read:
```scala
def nextDateRangeOption() = r.nextStringOption().map(mkRangeFn(toSQLDate))
```

**Comment by tminglei** [comment]
Oh, yes, thanks for pointing it out!

**Comment by tminglei** [comment]
Released v0.11.2 to include the fix. pls help check. _Thanks!_

**Comment by jeffreyr-porch** [comment]
v0.11.2 seems to be working for our use cases, and I have another team trying it for their use cases. I'll file a new ticket if we find any issues. Thanks!
